CRM/Contact/Form/Task/PDFLetterCommon.php

```diff
@@ -396,16 +410,53 @@ public static function postProcess(&$form) {
       $html[] = $tokenHtml;
     }
 
+    $saveToFile = count($activityIds) === 1 && Civi::settings()->get('recordGeneratedLetters') === 'combined-attached';
+    if ($saveToFile) {
+      // We need to capture data from various output-generators whose only
+      // consistent output format is writing to the console (for download).
+      $tmpFile = tempnam(sys_get_temp_dir(), 'PDFLetterCommon-');
+      $tmpFh = fopen($tmpFile, 'w');
+      ob_start(function ($buf) use ($tmpFh) {
+        // Pass to file, then continue to stdout.
+        fwrite($tmpFh, $buf);
+        return FALSE;
+      });
+    }
+
     $type = $formValues['document_type'];
+    $mimeType = self::getMimeType($type);
 
     if ($type == 'pdf') {
-      CRM_Utils_PDF_Utils::html2pdf($html, "CiviLetter.pdf", FALSE, $formValues);
+      $fileName = "CiviLetter.$type";
+      CRM_Utils_PDF_Utils::html2pdf($html, $fileName, FALSE, $formValues);
     }
     elseif (!empty($formValues['document_file_path'])) {
-      CRM_Utils_PDF_Document::printDocuments($formValues['document_file_path'], $html, $type, $zip);
+      $fileName = pathinfo($formValues['document_file_path'], PATHINFO_FILENAME) . '.' . $type;
+      CRM_Utils_PDF_Document::printDocuments($html, $fileName, $type, $zip);
     }
     else {
-      CRM_Utils_PDF_Document::html2doc($html, "CiviLetter.$type", $formValues);
+      $fileName = "CiviLetter.$type";
+      CRM_Utils_PDF_Document::html2doc($html, $fileName, $formValues);
+    }
+
+    if ($saveToFile) {
+      ob_end_flush();
+      fclose($tmpFh);
+      $content = file_get_contents($tmpFile, NULL, NULL, NULL, 5);
+      if (empty($content)) {
+        throw new \CRM_Core_Exception("Failed to capture document content (type=$type)!");
```

Comment:
No, they're not. It's not something I've had a strong opinion on, but I think there are a few reasons:
 * It's easier to search Google/StackExchange for errors if the textual representation is consistent.
 * For translators trying to prioritize/measure their progress, the exception messages have a poor cost/benefit.
 * `ts()` depends on some settings/services. Sometimes the exceptional circumstance affects/impedes `ts()`.